src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3431,32 +3432,34 @@ internal Buffer(IEnumerable<TElement> source, bool queryInterfaces = true)
                     if (collection != null)
                     {
                         count = collection.Count;
-                        if (count > 0)
+                        if (count != 0)
                         {
                             items = new TElement[count];
                             collection.CopyTo(items, 0);
                         }
+                        goto storeArray;
```

Comment:
You're completely correct. The only case where it saves anything but a single compare is in the case of a zero-item collection source, and there not only does it not save much, but that case has a much more straightforward way of dealing with it by just bailing out then, anyway.